Add documentation for resource customization (#1292)
istio-ecosystem/sail-operator: docs/general/resource-customization.adoc

link:../../README.adoc[Return to Project Root]

== Table of Contents

* <<custom-resource-annotations>>
** <<sailoperator-ignore-annotation>>
*** <<use-cases>>
*** <<how-it-works>>
*** <<usage>>
**** <<example-adding-annotation-to-webhook>>
*** <<removing-annotation>>
*** <<supported-resources>>
*** <<limitations>>

> *** <<implementation-details>>

Reviewer: ```suggestion
* <<custom-resource-annotations>>
** <<sailoperator-ignore-annotation>>
*** <<usage>>
**** <<example-adding-annotation-to-webhook>>
*** <<removing-annotation>>
*** <<supported-resources>>
*** <<limitations>>
```

Author: Modified

== Resource Customization

The Sail Operator supports customizing resources that it manages. This document describes the available customization options and how to use them.

[[custom-resource-annotations]]
=== Custom Resource Annotations

Special annotations can be added to resources to modify Sail Operator's behavior.

[[sailoperator-ignore-annotation]]
==== sailoperator.io/ignore Annotation

> The `sailoperator.io/ignore` annotation allows you to prevent the Sail Operator from reconciling changes to specific resources that it manages. This is useful when you need to make manual modifications to operator-managed resources that should not be automatically reverted.

Reviewer: ```suggestion
The `sailoperator.io/ignore` annotation allows you to prevent the Sail Operator from re-applying the helm charts in response to a change in one of the resources that it manages. This is useful when you or another controller needs to make modifications to operator-managed resources and the operator is continually overwriting the changes.

See these issues: https://github.com/istio-ecosystem/sail-operator/issues/430 and https://github.com/istio-ecosystem/sail-operator/issues/1148

This **does not** prevent the operator from re-applying the helm charts in response to other managed resources changing. 
```

Author: Modified

[[use-cases]]
===== Use Cases

The `sailoperator.io/ignore` annotation is helpful in the following scenarios (non-exhaustive list):

* **Customizing webhook configurations**: You need to modify webhook settings that should persist across reconciliation cycles
* **Adding labels or annotations**: You want to add custom labels or annotations to resources without them being removed
* **Tuning autoscaler settings**: You need to adjust `HorizontalPodAutoscaler` configurations based on observed behavior
* **Managing image pull secrets**: You want to add additional image pull secrets to `ServiceAccount` resources
* **Testing and debugging**: You need to temporarily modify a resource for troubleshooting purposes

> 

Reviewer: ```suggestion
```

Author: Removed

[[how-it-works]]
===== How It Works

Sail Operator re-applies all helm charts when it detects a change in any of the resources it deployed.
When the `sailoperator.io/ignore` annotation is set to `"true"` on a resource:

* **UPDATE event reconciliation is skipped** - The operator ignores update events for that specific resource, allowing manual modifications to persist
* **Global reconciliation still applies** - When other resources change, Sail Operator re-applies all Helm charts, including resources with this annotation. The `sailoperator.io/ignore: "true"` annotation still remains set on the resources.
* **Partial field protection** - Only fields defined in the Helm templates are reverted; additional annotations and labels you add will persist, but modifications to existing template-defined fields (such as labels or pod command-line arguments) will be reverted

[NOTE]
====
The annotation value must be exactly the string `"true"`. Any other value, such as `"True"`, `"yes"`, or `"1"`, will result in normal reconciliation behavior.

> ====

Reviewer: ```suggestion
```

Author: Don't you think the users should know these points?   
In the meantime, looking at Istio documentation they didn't expand on how the annotations work (eg. [authz-dry-run](https://istio.io/latest/docs/tasks/security/authorization/authz-dry-run/#summary))

Reviewer: These are covered in the intro section: https://github.com/istio-ecosystem/sail-operator/pull/1292#discussion_r3100393981. I don't think the implementation details matter for users. Just what problem the annotation is solving and how to use it.

Author: Removed

  sailoperator.io/ignore-
----

[NOTE]
====
After removing the annotation, the operator will reconcile the resource on the next UPDATE event, reverting any manual changes back to the state defined in the `Istio` or `IstioRevision` resource.
====

[[supported-resources]]
===== Supported Resources

The `sailoperator.io/ignore` annotation works on any resource that is:

* Owned by an `Istio` (has an `ownerReference` pointing to an `IstioRevision`)
* Subject to UPDATE events

Common resources include:

* `Deployment`
* `Service`
* `ServiceAccount`
* `ConfigMap`
* `MutatingWebhookConfiguration`
* `ValidatingWebhookConfiguration`
* `HorizontalPodAutoscaler`
* `PodDisruptionBudget`

[[limitations]]
===== Limitations and Considerations

* **Annotation is not preserved on recreation**: If the resource is deleted and recreated by the operator, the annotation will not be set automatically. You must re-apply it manually.
* **Manual maintenance required**: You are responsible for maintaining resources with this annotation. When upgrading Istio versions or when the parent `Istio` or `IstioRevision` resource changes, manual modifications will be reverted and may need to be re-applied.

> * **Configuration drift**: Using this annotation can lead to configuration drift between your intended state and the actual state. Use it sparingly and record which resources have this annotation in your documentation.

Reviewer: ```suggestion
* **Configuration drift**: Using this annotation can lead to configuration drift between your intended state and the actual state.
```

Author: Removed

[[implementation-details]]
===== Implementation Details

Here is the technical implementation:

* **Predicate Function**: `pkg/predicate/predicate.go:IgnoreUpdateWhenAnnotation()`
* **Integration Test**: `tests/integration/api/istiorevision_test.go`
* **Documentation**: This annotation was introduced in commit https://github.com/istio-ecosystem/sail-operator/commit/2543fb8[2543fb8]

> The implementation uses a controller-runtime predicate that filters UPDATE events based on the annotation value, preventing them from triggering reconciliation.

Reviewer: ```suggestion
```

Author: I'm ok to remove this

[[use-cases]]
===== Use Cases

The `sailoperator.io/ignore` annotation is helpful in the following scenarios (non-exhaustive list):

* **Customizing webhook configurations**: You need to modify webhook settings that should persist across reconciliation cycles
* **Adding labels or annotations**: You want to add custom labels or annotations to resources without them being removed
* **Tuning autoscaler settings**: You need to adjust `HorizontalPodAutoscaler` configurations based on observed behavior
* **Managing image pull secrets**: You want to add additional image pull secrets to `ServiceAccount` resources
* **Testing and debugging**: You need to temporarily modify a resource for troubleshooting purposes

[[how-it-works]]
===== How It Works

When the `sailoperator.io/ignore` annotation is set to `"true"` on a resource:

* The operator **will skip reconciling UPDATE events** for that resource
* The resource **must still be owned** by the `Istio` (the ownerReference remains the associated `IstioRevision`)

> * The operator **will not revert manual changes** made to resources with this annotation

Author: You're right ! Going to modify the doc to be more accurate how it works

[[use-cases]]
===== Use Cases

The `sailoperator.io/ignore` annotation is helpful in the following scenarios (non-exhaustive list):

* **Customizing webhook configurations**: You need to modify webhook settings that should persist across reconciliation cycles
* **Adding labels or annotations**: You want to add custom labels or annotations to resources without them being removed
* **Tuning autoscaler settings**: You need to adjust `HorizontalPodAutoscaler` configurations based on observed behavior
* **Managing image pull secrets**: You want to add additional image pull secrets to `ServiceAccount` resources
* **Testing and debugging**: You need to temporarily modify a resource for troubleshooting purposes

[[how-it-works]]
===== How It Works

Sail Operator re-applies all helm charts when it detects a change in any of the resources it deployed.  
When the `sailoperator.io/ignore` annotation is set to `"true"` on a resource:

> * **Ownership is preserved** - The resource must remain owned by its associated `Istio` (the ownerReference remains the associated `IstioRevision`)

Author: Removed, this is not necessary to be mentionned

* **UPDATE event reconciliation is skipped** - The operator ignores update events for that specific resource, allowing manual modifications to persist

> * **CREATE and DELETE events remain active** - The annotation only affects UPDATE events; creation and deletion are processed normally (ie. the reconciliation will be triggered and all the changes will be reverted back to the state defined in the `Istio` or `IstioRevision` resource)

Author: Right, I removed this confusing point

+
[source,bash]
----
kubectl get mutatingwebhookconfiguration \
  istio-sidecar-injector \
  -n istio-system \
  -o jsonpath='{.metadata.labels.app}'
----
+
Expected output:
+
[source,console]
----
sidecar-injector-test
----

[[removing-annotation]]
===== Removing the Annotation

To re-enable reconciliation for a resource, simply remove the annotation:

[source,bash]
----
kubectl annotate mutatingwebhookconfiguration \
  istio-sidecar-injector \
  -n istio-system \
  sailoperator.io/ignore-
----

[NOTE]
====
After removing the annotation, the operator will reconcile the resource on the next UPDATE event, reverting any manual changes back to the state defined in the `Istio` or `IstioRevision` resource.
====

[[supported-resources]]
===== Supported Resources

The `sailoperator.io/ignore` annotation works on any resource that is:

* Owned by an `Istio` (has an `ownerReference` pointing to an `IstioRevision`)
* Subject to UPDATE events

Common resources include:

* `Deployment`
* `Service`
* `ServiceAccount`
* `ConfigMap`
* `MutatingWebhookConfiguration`
* `ValidatingWebhookConfiguration`
* `HorizontalPodAutoscaler`
* `PodDisruptionBudget`

[[limitations]]

> ===== Limitations and Considerations

Author: I removed the duplicated points. IMO these remaining ones should be mentioned, WDYT?
